File: src/decoders/timetable-class-lesson.ts
import { SessionHandle, Subject, TimetableClassLesson } from "~/models";
import { decodeTimetableClass } from "./timetable-class";
import { decodeSubject } from "./subject";

export const decodeTimetableClassLesson = (item: any, session: SessionHandle): TimetableClassLesson => {
  const virtualClassrooms: string[] = [];
  const teacherNames: string[] = [];
  const personalNames: string[] = [];
  const classrooms: string[] = [];
  const groupNames: string[] = [];
  let subject: Subject | undefined;
  let lessonResourceID: string | undefined;

  if (item.listeVisios) {
    for (const virtualClassroom of item.listeVisios.V) {
      virtualClassrooms.push(virtualClassroom.url);
    }
  }

  if (item.ListeContenus) {
    for (const data of item.ListeContenus.V) {
      switch (data.G) {
        case 16: // Subject
          subject = decodeSubject(data);
          break;
        case 3: // Teacher
          teacherNames.push(data.L);
          break;
        case 34: // Personal
          personalNames.push(data.L);
          break;
        case 17: // Room
          classrooms.push(data.L);
          break;
        case 2: // Group
          groupNames.push(data.L);
          break;
      }
    }
  }

  if (item.AvecCdT && item.cahierDeTextes) {
    lessonResourceID = item.cahierDeTextes.V.N;
  }

  return {
    is: "lesson",
    ...decodeTimetableClass(item, session),
    kind: item.G,
    status: item.Statut,
    canceled: item.estAnnule ?? false,
    exempted: item.dispenseEleve ?? false,
    test: item.cahierDeTextes?.V.estDevoir ?? false,
    virtualClassrooms,
    personalNames,
    teacherNames,
    classrooms,
    groupNames,
    subject,
    lessonResourceID
  };
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/decoders/timetable-class-lesson.kt
package ink.literate.pawnote.decoders

import ink.literate.pawnote.models.Subject
import ink.literate.pawnote.models.TimetableClassLesson
import kotlinx.serialization.json.*

fun decodeTimetableClassLesson(item: JsonObject): TimetableClassLesson {
  val virtualClassrooms: MutableList<String> = mutableListOf()
  val teacherNames: MutableList<String> = mutableListOf()
  val personalNames: MutableList<String> = mutableListOf()
  val classrooms: MutableList<String> = mutableListOf()
  val groupNames: MutableList<String> = mutableListOf()

  var subject: Subject? = null
  var lessonResourceID: String? = null

  if (item.containsKey("listeVisios")) {
    for (virtualClassroom in item["listeVisios"]!!.jsonObject["V"]!!.jsonArray) virtualClassrooms
        .add(virtualClassroom.jsonObject["url"]!!.jsonPrimitive.content)
  }

  if (item.containsKey("ListeContenus")) {
    for (data in item["ListeContenus"]!!.jsonObject["V"]!!.jsonArray) {
      val dataObj = data.jsonObject
      when (dataObj["G"]!!.jsonPrimitive.int) {
        16 -> subject = decodeSubject(dataObj)
        3 -> teacherNames.add(dataObj["L"]!!.jsonPrimitive.content)
        34 -> personalNames.add(dataObj["L"]!!.jsonPrimitive.content)
        17 -> classrooms.add(dataObj["L"]!!.jsonPrimitive.content)
        2 -> groupNames.add(dataObj["L"]!!.jsonPrimitive.content)
      }
    }
  }

  if (item["AvecCdT"]?.jsonPrimitive?.boolean == true && item.containsKey("cahierDeTextes"))
      lessonResourceID =
          item["cahierDeTextes"]!!.jsonObject["V"]!!.jsonObject["N"]!!.jsonPrimitive.content

  return TimetableClassLesson(
      kind = item["G"]!!.jsonPrimitive.int,
      status = item["Statut"]?.jsonPrimitive?.content,
      canceled = item["estAnnule"]?.jsonPrimitive?.boolean ?: false,
      exempted = item["dispenseEleve"]?.jsonPrimitive?.boolean ?: false,
      test =
          if (item["cahierDeTextes"] != null)
              item["cahierDeTextes"]!!
                  .jsonObject["V"]!!
                  .jsonObject["estDevoir"]
                  ?.jsonPrimitive
                  ?.boolean ?: false
          else false,
      virtualClassrooms = virtualClassrooms,
      personalNames = personalNames,
      teacherNames = teacherNames,
      classrooms = classrooms,
      groupNames = groupNames,
      subject = subject,
      lessonResourceID = lessonResourceID)
}
